### DISCUSSION TITLE
### Improve support for ASP.NET 3.1, 5, and 6.

### DISCUSSION THREAD

**Discussion by mconnew** [CoreWCF/CoreWCF#603] [General]
### Improve support for ASP.NET 3.1, 5, and 6.

ASP.NET 2.1 has reached end of support. Drop support for ASP.NET 2.1 and improve support for ASP.NET 3.1, 5, and especially 6.

For example, these backward compatibility settings shouldn't be required when using CoreWCF in a new ASP.NET project:

```
builder.Services.Configure<IISServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
builder.Services.Configure<KestrelServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
```

_Originally posted by @mvarblow in https://github.com/CoreWCF/CoreWCF/issues/234#issuecomment-1005125539_

**Comment by mconnew** [comment]
@mvarblow, I've moved this to a discussion and will be deleting the comment in the feature roadmap issue as it's not a feature, but you raise a few things which are worth commenting on. ASP.NET 2.1 has not reached end of support, only .NET Core 2.1. You may not be aware, but ASP.NET2.1 also runs on .NET Framework, and it's how we support .NET Framework in CoreWCF. Once the 1.0 release is completed (in the coming weeks), I'll start work on the next version (2.0) which will have a minimum requirement of .NET 6.0. The 1.0 release will be in indefinite support while there is still demand for running in .NET Framework. The target scenario is as a stepping stone on the way to migrating a large app to .NET. You can move your service to CoreWCF and then incrementally work to migrate the code base across to netstandard2.0, then once that's all done, you can flip the switch and run on .NET.  

As for the AllowSynchronousIO issue, there are a couple of places where things can be improved to remove the need for most use cases, but it's impossible to make it so it's never needed. The issue is that there are many scenarios where the serializer (DataContractSerializer and XmlSerializer) writes directly to the output stream. Without async serialization support, the serializers will only write to the output using synchronous api's. There is an issue in the runtime repo requesting support for asynchronous serialization. Once that is available, we can remove the need for setting that property. In the 2.0 release, I'll definitely do what I can to reduce the number of scenarios where setting that is needed.  

If you have any more questions about either of these, feel free to ask in this discussion thread.

**Reply by mvarblow** [reply]
Thanks for the explanation. This makes perfect sense and seems like a great approach.  And I see that I should have gathered this from other posts. Thanks for taking the time to give a detailed response.

I really appreciate the great work you're doing to bring WCF services to new .NET runtimes and improve on the old framework in many ways.

**Reply by mconnew** [reply]
@mvarblow, I have a PR out now which will eliminate the need to use AllowSynchronousIO for most use cases. You can find it at #648

**Reply by mvarblow** [reply]
Nice!

**Comment by a-shoemaker** [comment]
I'm wondering if there is a feeling on when 2.0 will be released.  Given the dependency on out of support / deprecated packages, I'm hoping really soon?  Or is the expectation that one would just go out and create their own packages addressing this for compliance in the meantime (e.g., #766)?

**Reply by mconnew** [reply]
CoreWCF doesn't depend on any out of support of deprecated packages. It does have a dependency on some asp.net core packages to provide support for .NET Framework, and they are fully supported for .NET Framework. When running on .NET 6 or later, if you either reference the Web sdk with a project line of `<Project Sdk="Microsoft.NET.Sdk.Web">` or add the following to your csproj:
```xml
  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>
```
Then it will use the in-box version of asp.net core that ships with .NET 6 (or any later version of .NET that you are using).  

Are there any compliance issues that aren't addressed by this?

**Reply by a-shoemaker** [reply]
So if I set it up right (which perhaps is my real issue), then say Microsoft.AspNetCore.Hosting.Abstractions (deprecated package, last version is 2.2.0) will not come into play then at all.  Or is it perhaps alternatively should not consider that package out of support per say?

(And thank you greatly for your fast reply!)

**Reply by a-shoemaker** [reply]
Apologies, I got it tested out and I did have an issue with my setup.  Your comment fully satisfies my initial question.  Thank you for your quick and thoughtful response.

**Reply by mconnew** [reply]
The Microsoft.AspNetCore.* nuget packages with version 2.2 are out of support, but the 2.1 versions are still in support for .NET Framework only.
